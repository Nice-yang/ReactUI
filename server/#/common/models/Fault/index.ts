export type FaultArguments<T> = Parameters<
  (message: string, data?: T, level?: number) => void
>;

export class Fault<T extends object = any> {
  public message: string;
  public data?: T;
  public level: number;
  public stack?: string;

  constructor(...args: FaultArguments<T>) {
    this.message = args[0];
    this.data = args[1];
    this.level = args[2] || 0;
    this.stack = new Error().stack;
  }
}
