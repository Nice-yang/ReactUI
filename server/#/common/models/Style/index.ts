export type StyleObject = Array<{ key: string; value: string }>;
