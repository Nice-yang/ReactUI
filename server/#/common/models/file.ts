export interface TagCursor {
  fileName: string;
  lineNumber: number;
  columnNumber: number;
}
