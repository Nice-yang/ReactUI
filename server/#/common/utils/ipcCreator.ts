import * as Electron from 'electron';
import * as ipc from 'electron-better-ipc';

export function mainToRendererRequest<T, P>(channel: string) {
  const callRenderer = (win: Electron.BrowserWindow, arg: T) =>
    ipc.callRenderer && ipc.callRenderer(win, channel, arg);
  const answerMain = (cb: (arg: T) => Promise<P>) =>
    ipc.answerMain && ipc.answerMain(channel, cb);

  return {
    callRenderer,
    answerMain,
  };
}
