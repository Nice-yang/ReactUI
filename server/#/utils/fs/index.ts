import * as fs from 'fs';
import { bindNodeCallback } from 'rxjs';

export const writeFile = bindNodeCallback(fs.writeFile);
