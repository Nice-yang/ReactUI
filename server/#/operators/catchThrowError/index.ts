import { Fault, FaultArguments } from '#/common/models/Fault';
import { Observable } from 'rxjs';

export const catchThrowFault = <T>(...args: FaultArguments<any>) => (
  source: Observable<T>,
) => {
  return new Observable<T>(observer => {
    return source.subscribe({
      next(x) {
        observer.next(x);
      },
      error(err) {
        observer.error(
          new Fault(args[0], { ...args[1], baseError: err }, args[2]),
        );
      },
      complete() {
        observer.complete();
      },
    });
  });
};
