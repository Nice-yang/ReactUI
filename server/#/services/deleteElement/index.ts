import { Fault } from '#/common/models/Fault';
import { TagCursor } from '#/common/models/file';
import { ReplacementBuilder } from '#/utils/ReplacementBuilder';
import { findNode$, isAtCursor } from '#/utils/tsNode';
import { createSourceFileFromText } from '#/utils/tsSourceFile';
import * as R from 'ramda';
import { of, throwError } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import * as ts from 'typescript';

const deleteElement = R.curry((cursor: TagCursor, fileContent: string) => {
  return of(createSourceFileFromText(cursor.fileName, fileContent)).pipe(
    switchMap(sourceFileNode =>
      findNode$<ts.JsxElement>(R.both(ts.isJsxElement, isAtCursor(cursor)))(
        sourceFileNode,
      ).pipe(
        catchError(err =>
          throwError(
            new Fault('Element not found at cursor', {
              cursorLocation: cursor,
              baseError: err,
            }),
          ),
        ),
        map(elementNode => {
          const replacementBuilder = new ReplacementBuilder(sourceFileNode);
          replacementBuilder.deleteNode(elementNode);

          if (
            !ts.isJsxElement(elementNode.parent) &&
            !ts.isJsxFragment(elementNode.parent)
          ) {
            replacementBuilder.insert(elementNode.getStart(), 'null');
          }

          return replacementBuilder.applyReplacements();
        }),
      ),
    ),
  );
});

export default deleteElement;
