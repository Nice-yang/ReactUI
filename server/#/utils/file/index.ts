import { Fault } from '#/common/models/Fault';
import { readFile, writeFile } from 'fs';
import * as R from 'ramda';
import { bindNodeCallback, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';

export const readFile$ = bindNodeCallback(readFile);
export const writeFile$ = bindNodeCallback(writeFile);
export const readFileToString$ = (fileName: string) =>
  readFile$(fileName).pipe(
    catchError(err =>
      throwError(
        new Fault('Cannot read file', { error: JSON.stringify(err), fileName }),
      ),
    ),
    map(R.toString),
  );
