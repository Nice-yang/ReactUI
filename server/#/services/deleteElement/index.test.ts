import { Fault } from '#/common/models/Fault';
import { readFileToString$ } from '#/utils/file';
import { switchMap } from 'rxjs/operators';
import deleteElement from './index';

it('deletes child element', done => {
  const fileName = `${__dirname}/__fixtures__/1.tsx`;
  readFileToString$(fileName)
    .pipe(
      switchMap(deleteElement({ lineNumber: 5, columnNumber: 5, fileName })),
    )
    .subscribe(x => {
      expect(x).toMatchSnapshot();
      done();
    });
});

it('deletes element with no parent', done => {
  const fileName = `${__dirname}/__fixtures__/replaceWithNull1.tsx`;
  readFileToString$(fileName)
    .pipe(
      switchMap(deleteElement({ lineNumber: 3, columnNumber: 25, fileName })),
    )
    .subscribe(x => {
      expect(x).toMatchSnapshot();
      done();
    });
});

it('throws Fault if element at cursor is not found', done => {
  const fileName = `${__dirname}/__fixtures__/1.tsx`;

  readFileToString$(fileName)
    .pipe(
      switchMap(deleteElement({ lineNumber: 3, columnNumber: 7, fileName })),
    )
    .subscribe({
      error: x => {
        expect(x instanceof Fault).toBeTruthy();
        expect(x).toMatchSnapshot();
        done();
      },
    });
});
