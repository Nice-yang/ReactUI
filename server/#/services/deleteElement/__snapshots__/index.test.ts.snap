// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deletes child element 1`] = `
"import * as React from 'react';

const Component = () => (
  <div>
    
  </div>
);

export default Component;
"
`;

exports[`deletes element with no parent 1`] = `
"import * as React from 'react';

const Component = () => null;

export default Component;
"
`;

exports[`throws Fault if element at cursor is not found 1`] = `
Fault {
  "data": Object {
    "baseError": Fault {
      "data": Object {
        "rootNode": SourceFileObject {
          "amdDependencies": Array [],
          "bindDiagnostics": Array [],
          "bindSuggestionDiagnostics": undefined,
          "checkJsDirective": undefined,
          "end": 136,
          "endOfFileToken": TokenObject {
            "end": 136,
            "flags": 0,
            "kind": 1,
            "parent": [Circular],
            "pos": 135,
          },
          "externalModuleIndicator": NodeObject {
            "decorators": undefined,
            "end": 31,
            "flags": 0,
            "importClause": NodeObject {
              "end": 17,
              "flags": 0,
              "kind": 250,
              "namedBindings": NodeObject {
                "end": 17,
                "flags": 0,
                "kind": 251,
                "name": IdentifierObject {
                  "end": 17,
                  "escapedText": "React",
                  "flags": 0,
                  "parent": [Circular],
                  "pos": 11,
                },
                "parent": [Circular],
                "pos": 6,
                "transformFlags": undefined,
              },
              "parent": [Circular],
              "pos": 6,
              "transformFlags": undefined,
            },
            "kind": 249,
            "modifierFlagsCache": 536870912,
            "modifiers": undefined,
            "moduleSpecifier": TokenObject {
              "end": 30,
              "flags": 0,
              "kind": 10,
              "parent": [Circular],
              "pos": 22,
              "text": "react",
            },
            "parent": [Circular],
            "pos": 0,
            "transformFlags": undefined,
          },
          "fileName": "/Users/I0459/scratch/personal/beragi/server/#/services/deleteElement/__fixtures__/1.tsx",
          "flags": 0,
          "hasNoDefaultLib": false,
          "identifierCount": 7,
          "identifiers": Map {
            "React" => "React",
            "react" => "react",
            "Component" => "Component",
            "div" => "div",
            "button" => "button",
          },
          "isDeclarationFile": false,
          "kind": 279,
          "languageVariant": 1,
          "languageVersion": 6,
          "libReferenceDirectives": Array [],
          "lineMap": Array [
            0,
            32,
            33,
            59,
            67,
            97,
            106,
            109,
            110,
            136,
          ],
          "nodeCount": 31,
          "parent": undefined,
          "parseDiagnostics": Array [],
          "pos": 0,
          "pragmas": Map {},
          "referencedFiles": Array [],
          "scriptKind": 4,
          "statements": Array [
            NodeObject {
              "decorators": undefined,
              "end": 31,
              "flags": 0,
              "importClause": NodeObject {
                "end": 17,
                "flags": 0,
                "kind": 250,
                "namedBindings": NodeObject {
                  "end": 17,
                  "flags": 0,
                  "kind": 251,
                  "name": IdentifierObject {
                    "end": 17,
                    "escapedText": "React",
                    "flags": 0,
                    "parent": [Circular],
                    "pos": 11,
                  },
                  "parent": [Circular],
                  "pos": 6,
                  "transformFlags": undefined,
                },
                "parent": [Circular],
                "pos": 6,
                "transformFlags": undefined,
              },
              "kind": 249,
              "modifierFlagsCache": 536870912,
              "modifiers": undefined,
              "moduleSpecifier": TokenObject {
                "end": 30,
                "flags": 0,
                "kind": 10,
                "parent": [Circular],
                "pos": 22,
                "text": "react",
              },
              "parent": [Circular],
              "pos": 0,
              "transformFlags": undefined,
            },
            NodeObject {
              "declarationList": NodeObject {
                "declarations": Array [
                  NodeObject {
                    "end": 107,
                    "flags": 0,
                    "initializer": NodeObject {
                      "body": NodeObject {
                        "end": 107,
                        "expression": NodeObject {
                          "children": Array [
                            TokenObject {
                              "containsOnlyWhiteSpaces": true,
                              "end": 71,
                              "flags": 0,
                              "kind": 11,
                              "parent": [Circular],
                              "pos": 66,
                            },
                            NodeObject {
                              "children": Array [
                                TokenObject {
                                  "containsOnlyWhiteSpaces": false,
                                  "end": 87,
                                  "flags": 0,
                                  "kind": 11,
                                  "parent": [Circular],
                                  "pos": 79,
                                },
                              ],
                              "closingElement": NodeObject {
                                "end": 96,
                                "flags": 0,
                                "kind": 263,
                                "parent": [Circular],
                                "pos": 87,
                                "tagName": IdentifierObject {
                                  "end": 95,
                                  "escapedText": "button",
                                  "flags": 0,
                                  "parent": [Circular],
                                  "pos": 89,
                                },
                                "transformFlags": undefined,
                              },
                              "end": 96,
                              "flags": 0,
                              "kind": 260,
                              "openingElement": NodeObject {
                                "attributes": NodeObject {
                                  "end": 78,
                                  "flags": 0,
                                  "kind": 268,
                                  "parent": [Circular],
                                  "pos": 78,
                                  "properties": Array [],
                                  "transformFlags": undefined,
                                },
                                "end": 79,
                                "flags": 0,
                                "kind": 262,
                                "parent": [Circular],
                                "pos": 71,
                                "tagName": IdentifierObject {
                                  "end": 78,
                                  "escapedText": "button",
                                  "flags": 0,
                                  "parent": [Circular],
                                  "pos": 72,
                                },
                                "transformFlags": undefined,
                                "typeArguments": undefined,
                              },
                              "parent": [Circular],
                              "pos": 71,
                              "transformFlags": undefined,
                            },
                            TokenObject {
                              "containsOnlyWhiteSpaces": true,
                              "end": 99,
                              "flags": 0,
                              "kind": 11,
                              "parent": [Circular],
                              "pos": 96,
                            },
                          ],
                          "closingElement": NodeObject {
                            "end": 105,
                            "flags": 0,
                            "kind": 263,
                            "parent": [Circular],
                            "pos": 99,
                            "tagName": IdentifierObject {
                              "end": 104,
                              "escapedText": "div",
                              "flags": 0,
                              "parent": [Circular],
                              "pos": 101,
                            },
                            "transformFlags": undefined,
                          },
                          "end": 105,
                          "flags": 0,
                          "kind": 260,
                          "openingElement": NodeObject {
                            "attributes": NodeObject {
                              "end": 65,
                              "flags": 0,
                              "kind": 268,
                              "parent": [Circular],
                              "pos": 65,
                              "properties": Array [],
                              "transformFlags": undefined,
                            },
                            "end": 66,
                            "flags": 0,
                            "kind": 262,
                            "parent": [Circular],
                            "pos": 58,
                            "tagName": IdentifierObject {
                              "end": 65,
                              "escapedText": "div",
                              "flags": 0,
                              "parent": [Circular],
                              "pos": 62,
                            },
                            "transformFlags": undefined,
                            "typeArguments": undefined,
                          },
                          "parent": [Circular],
                          "pos": 58,
                          "transformFlags": undefined,
                        },
                        "flags": 0,
                        "kind": 195,
                        "parent": [Circular],
                        "pos": 56,
                        "transformFlags": undefined,
                      },
                      "end": 107,
                      "equalsGreaterThanToken": TokenObject {
                        "end": 56,
                        "flags": 0,
                        "kind": 37,
                        "parent": [Circular],
                        "pos": 53,
                      },
                      "flags": 0,
                      "kind": 197,
                      "modifierFlagsCache": 536870912,
                      "modifiers": undefined,
                      "parameters": Array [],
                      "parent": [Circular],
                      "pos": 50,
                      "transformFlags": undefined,
                      "typeParameters": undefined,
                    },
                    "kind": 237,
                    "name": IdentifierObject {
                      "end": 48,
                      "escapedText": "Component",
                      "flags": 0,
                      "parent": [Circular],
                      "pos": 38,
                    },
                    "parent": [Circular],
                    "pos": 38,
                    "transformFlags": undefined,
                    "type": undefined,
                  },
                ],
                "end": 107,
                "flags": 2,
                "kind": 238,
                "parent": [Circular],
                "pos": 31,
                "transformFlags": undefined,
              },
              "decorators": undefined,
              "end": 108,
              "flags": 0,
              "kind": 219,
              "modifiers": undefined,
              "parent": [Circular],
              "pos": 31,
              "transformFlags": undefined,
            },
            NodeObject {
              "decorators": undefined,
              "end": 135,
              "expression": IdentifierObject {
                "end": 134,
                "escapedText": "Component",
                "flags": 0,
                "parent": [Circular],
                "pos": 124,
              },
              "flags": 0,
              "kind": 254,
              "modifiers": undefined,
              "parent": [Circular],
              "pos": 108,
              "transformFlags": undefined,
            },
          ],
          "text": "import * as React from 'react';

const Component = () => (
  <div>
    <button>Click me</button>
  </div>
);

export default Component;
",
          "transformFlags": undefined,
          "typeReferenceDirectives": Array [],
        },
      },
      "level": 0,
      "message": "Node not found",
      "stack": "Error
    at new Fault (/Users/I0459/scratch/personal/beragi/server/#/common/models/Fault/index.ts:11:18)
    at Observable.rxjs_1.Observable.subscriber [as _subscribe] (/Users/I0459/scratch/personal/beragi/server/#/utils/tsNode/index.ts:55:22)
    at Observable.Object.<anonymous>.Observable._trySubscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:231:19)
    at Observable.Object.<anonymous>.Observable.subscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:212:14)
    at CatchOperator.Object.<anonymous>.CatchOperator.call (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/operators/catchError.ts:107:19)
    at Observable.Object.<anonymous>.Observable.subscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:207:25)
    at MapOperator.Object.<anonymous>.MapOperator.call (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/operators/map.ts:59:19)
    at Observable.Object.<anonymous>.Observable.subscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:207:25)
    at /Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/util/subscribeTo.ts:22:23
    at Object.subscribeToResult (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/util/subscribeToResult.ts:25:29)",
    },
    "cursorLocation": Object {
      "columnNumber": 7,
      "fileName": "/Users/I0459/scratch/personal/beragi/server/#/services/deleteElement/__fixtures__/1.tsx",
      "lineNumber": 3,
    },
  },
  "level": 0,
  "message": "Element not found at cursor",
  "stack": "Error
    at new Fault (/Users/I0459/scratch/personal/beragi/server/#/common/models/Fault/index.ts:11:18)
    at CatchSubscriber.tsNode_1.findNode$.pipe.operators_1.catchError.err [as selector] (/Users/I0459/scratch/personal/beragi/server/#/services/deleteElement/index.ts:19:13)
    at CatchSubscriber.Object.<anonymous>.CatchSubscriber.error (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/operators/catchError.ts:132:23)
    at Observable.rxjs_1.Observable.subscriber [as _subscribe] (/Users/I0459/scratch/personal/beragi/server/#/utils/tsNode/index.ts:55:16)
    at Observable.Object.<anonymous>.Observable._trySubscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:231:19)
    at Observable.Object.<anonymous>.Observable.subscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:212:14)
    at CatchOperator.Object.<anonymous>.CatchOperator.call (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/operators/catchError.ts:107:19)
    at Observable.Object.<anonymous>.Observable.subscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:207:25)
    at MapOperator.Object.<anonymous>.MapOperator.call (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/operators/map.ts:59:19)
    at Observable.Object.<anonymous>.Observable.subscribe (/Users/I0459/scratch/personal/beragi/server/node_modules/rxjs/src/internal/Observable.ts:207:25)",
}
`;
